pkg/controller/volume/persistentvolume/scheduler_binder_test.go

```diff
@@ -1275,60 +1404,70 @@ func TestCheckBindings(t *testing.T) {
 }
 
 func TestBindPodVolumes(t *testing.T) {
-	scenarios := map[string]struct {
+	type scenarioType struct {
 		// Inputs
-		// These tests only support a single pv and pvc and static binding
 		bindingsNil bool // Pass in nil bindings slice
-		binding     *bindingInfo
-		cachedPV    *v1.PersistentVolume
-		cachedPVC   *v1.PersistentVolumeClaim
-		apiPV       *v1.PersistentVolume
+
+		// before assume
+		initPV  *v1.PersistentVolume
+		initPVC *v1.PersistentVolumeClaim
+
+		// assume PV & PVC with these binding results
+		binding          *bindingInfo
+		claimToProvision *v1.PersistentVolumeClaim
+
+		// API updates after assume before bind
+		apiPV  *v1.PersistentVolume
+		apiPVC *v1.PersistentVolumeClaim
 
 		// This function runs with a delay of 5 seconds
-		delayFunc func(*testing.T, *testEnv, *v1.Pod, *v1.PersistentVolume, *v1.PersistentVolumeClaim)
+		delayFunc func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim)
 
 		// Expected return values
 		shouldFail bool
-	}{
+	}
+
+	scenarios := map[string]scenarioType{
 		"nothing-to-bind-nil": {
 			bindingsNil: true,
 			shouldFail:  true,
 		},
 		"nothing-to-bind-empty": {},
 		"already-bound": {
-			binding:   binding1aBound,
-			cachedPV:  pvNode1aBound,
-			cachedPVC: boundPVCNode1a,
-		},
-		"binding-succeeds-after-time": {
-			binding:   binding1aBound,
-			cachedPV:  pvNode1a,
-			cachedPVC: unboundPVC,
+			initPV:     pvNode1aBound,
+			initPVC:    unboundPVC,
+			binding:    makeBinding(unboundPVC, pvNode1aBound),
+			apiPV:      pvNode1aBound,
+			apiPVC:     boundPVCNode1a,
+			shouldFail: true,
```

Comment:
If PV is already bound, BindPodVolumes actually will fail in bindAPIUpdate because object version conflict.

Reply:
Sorry, I misunderstood 'already-bound' test.  There are more cases:

Case 1: Before FindPodVolumes, PV is bound to PVC, PVC is not fully bound to PV

IIUC, this is original test case. BindPodVolumes will succeed after PVC is fully bound to PV by PV controller. I implement this behavior by using asyncFunc.

Case 2: Before FindPodVolumes, PV is not bound to PVC, PVC is not fully bound to PV

If PV is bound to PVC by PV controller before BindPodVolumes, BindPodVolumes will fail because API conflict.

I've added a test `bound-by-pv-controller-before-bind` for this.

Case 3: Before FindPodVolumes, PV is bound to PVC, PVC is fully bound to PV

BindPodVolumes will not be called. So we do not need test this.